onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,441 @@
+    ORT_ENFORCE(b_size_t.Ref() > (num_values * ml_data_type_->Size()),
+                "Values size must be less than total buffer size");
+    auto data_ptr = IAllocator::MakeUniquePtr<void>(allocator_, b_size_t);
+    ORT_RETURN_IF_NOT(data_ptr != nullptr, "SparseTensor Allocation failed for size: ", buffer_size);
+    if (IsDataTypeString()) {
+      // We own the buffer, so we must properly construct strings. Neither of the Tensors
+      // we construct on top of the buffer own it. We are constructing empty strings, hopefully
+      // nothrow and no buffer allocation
+      utils::ConstructStrings(data_ptr.get(), static_cast<int64_t>(num_values));
+    }
+    p_data_ = data_ptr.release();
+  }
+  buffer_size_ = buffer_size;
+  return Status::OK();
+}
+
+void SparseTensor::ReleaseBuffer() {
+  if (allocator_ && p_data_ != nullptr) {
+    // if current tensor is responsible for deleting the buffer
+    // and it is a string tensor, need to explicitly call string(s)
+    // __dtor(s).
+    utils::ReleaseTensorBuffer(allocator_, IsDataTypeString(), p_data_, values_.Shape().Size());
+  }
+  p_data_ = nullptr;
+  buffer_size_ = 0;
+}
+
+void SparseTensor::CopyStrings(const Tensor& src, Tensor& dst) const {
+  auto src_span = src.DataAsSpan<std::string>();
+  auto* dst_iter = dst.MutableData<std::string>();
+  std::copy(src_span.cbegin(), src_span.cend(), dst_iter);
+}
+
+Status SparseTensor::CopyData(const IDataTransfer& data_transfer,
+                              const std::vector<std::reference_wrapper<const Tensor>>& src,
+                              const std::vector<std::reference_wrapper<Tensor>>& dst) {
+  ORT_RETURN_IF_NOT(src.size() == dst.size(), "Must have the same size");
+  for (size_t i = 0; i < src.size(); ++i) {
+    const Tensor& src_t = src[i];
+    Tensor& dst_t = dst[i];
+    if (src_t.IsDataTypeString()) {
+      CopyStrings(src_t, dst_t);
+    } else {
+      ORT_RETURN_IF_ERROR(data_transfer.CopyTensor(src_t, dst_t));
+    }
+  }
+  return Status::OK();
+}
+
+SparseTensor::CooView SparseTensor::AsCoo() const {
+  ORT_ENFORCE(Format() == SparseFormat::kCoo, "Must contain Coo format");
+  ORT_ENFORCE(format_data_.size() == 1U, "Expecting one index");
+  return CooView(format_data_[0]);
+}
+
+std::vector<int64_t> SparseTensor::GetCooIndexDims(size_t values_count, size_t index_size) const {
+  std::vector<int64_t> index_dims{gsl::narrow<int64_t>(values_count)};
+  if (values_count * 2 == index_size) {
+    // Coor index
+    index_dims.push_back(2);
+  } else {
+    ORT_ENFORCE(values_count == index_size,
+                      "Index size must be equal to or twice the values size");
+  }
+  return index_dims;
+}
+
+void SparseTensor::InitCooIndex(const TensorShape& index_shape, const int64_t* index_data) {
+  format_data_.resize(1);
+  format_data_[0] = Tensor(DataTypeImpl::GetType<int64_t>(), index_shape, const_cast<int64_t*>(index_data), Location());
+  format_ = SparseFormat::kCoo;
```

Comment:
I think this should take `int64_t*` as input so the const_cast isn't hidden. Tensor provides mutable access to the data it contains, so you're somewhat lying about what could happen by saying index_data is const.

Reply:
Alternatively can format_data_ hold `const Tensor` values instead? That way we can justify the internal const_cast by storing the result in a Tensor instance that is immutable.

Reply:
SparseTensor class is immutable. I think this should be good enough.

Reply:
I made Use*() interfaces take non-const data.
I can not justify the expense of creating a new class that seemly doing the same thing as std::reference_wrapper<const Tensor> which is what we expose in the view classes anyway. We do not expose any mutating interfaces.

Reply:
with that change the Init* methods don't need to take a const* and do const_cast do they?

Reply:
will change to non-const